### PR TITLE
refactor: ante fee

### PR DESCRIPTION
**Pull request by JoowonYun**
## Effects

- [ ] Soft update
- [ ] Breaking change
- [ ] Chain fork related

## Major Reviewer

@green-blue0 @yoosah @psy2848048 

## Background

- When https://github.com/xpladev/xpla/pull/57 is merged, it will go to feature/xatp

## Summary

- Responding when there are multiple native tokens and XATP

## Checklist

- [ ] Backward compatible?
- [ ] Test enough in your local environment?
- [ ] Add related test cases?

### CODE REVIEW

File: ante/fee.go [hunk 1]
```diff
@@ -186,85 +175,59 @@ func (dfd DeductFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bo
...
+			if err != nil {
+				return ctx, err
+			}
+		}
 
-				err = DeductFees(dfd.bankKeeper, ctx, deductFeesFromAcc0, xplaFee)
-				if err != nil {
-					return ctx, err
-				}
+		if !xatpFees.Empty() {
```
Comment (psy2848048) [inline ante/fee.go:219]: XATP only exeutes on the node which received the Tx

I added a debug print line as https://github.com/xpladev/xpla/commit/0e53329a7c705b17f186cf10cd3eff83c9529b8e
(one common line vs only XATP Tx executes)

In the integration tests, all tests with the common fee prints the common print
But in the case of XATP tx, the tx is only executed in the node which is received the tx


Printed the event and it executes well

```
events:
_contract_address : xpla14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9s525s0h
_contract_address : xpla14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9s525s0h
action : transfer
from : xpla1z2k85n48ydfvzslrugwzl4j2u7vtdyf3xvucmc
to : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
amount : 360000
spender : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
amount : 1000000000000000000axpla
receiver : xpla17xpfvakm2amg962yls6f84z3kell8c5lw7muqw
amount : 1000000000000000000axpla
recipient : xpla17xpfvakm2amg962yls6f84z3kell8c5lw7muqw
sender : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
amount : 1000000000000000000axpla
sender : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
```

File: ante/fee.go [hunk 2]
```diff
@@ -186,85 +175,59 @@ func (dfd DeductFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bo
...
+			ratioDec, err := dfd.xatpKeeper.GetFeeInfoFromXATP(ctx, xatp.Denom)
+			if err != nil {
+				return ctx, err
+			}
 
-				deductFeesFromAcc0 := dfd.ak.GetAccount(ctx, XatpPayerAcc)
+			feeAmount := sdk.NewDecFromIntWithPrec(fee.Amount, int64(xatp.Decimals))
+			defaultFeeAmountDec := feeAmount.Quo(ratioDec)
+			defaultFeeAmount := defaultFeeAmountDec.Mul(sdk.NewDecWithPrec(10, xplatypes.DefaultDenomPrecision)).Ceil().BigInt()
+			xatpFees = xatpFees.Add(sdk.NewCoin(xplatypes.DefaultDenom, sdk.NewIntFromBigInt(defaultFeeAmount)))
```
Comment (psy2848048) [inline ante/fee.go:209]: The tests executes the native & CW20 transfer and I rated different gas limit
But the module withdraw same XPLA from XATP payer

Native transfer via bank module
The amount should be near 160000_000000000000 axpla (0.16 XPLA)
```
events:
_contract_address : xpla14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9s525s0h
_contract_address : xpla14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9s525s0h
action : transfer
from : xpla1z2k85n48ydfvzslrugwzl4j2u7vtdyf3xvucmc
to : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
amount : 160000
spender : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
amount : 1000000000000000000axpla
receiver : xpla17xpfvakm2amg962yls6f84z3kell8c5lw7muqw
amount : 1000000000000000000axpla
recipient : xpla17xpfvakm2amg962yls6f84z3kell8c5lw7muqw
sender : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
amount : 1000000000000000000axpla
sender : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
```

CW20 transfer
The amount should be near 405000_000000000000 axpla (0.405 XPLA)
```
_contract_address : xpla14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9s525s0h
_contract_address : xpla14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9s525s0h
action : transfer
from : xpla1u27snswkjpenlscgvszcfjmz8uy2y5qacx0826
to : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
amount : 405000
spender : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
amount : 1000000000000000000axpla
receiver : xpla17xpfvakm2amg962yls6f84z3kell8c5lw7muqw
amount : 1000000000000000000axpla
recipient : xpla17xpfvakm2amg962yls6f84z3kell8c5lw7muqw
sender : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
amount : 1000000000000000000axpla
sender : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
```

**Review by psy2848048 — COMMENTED** [review]
🙏

File: ante/fee.go (hunk 1 above)
Comment (psy2848048) [reply ante/fee.go:219]: Classified as a bad transaction in ante/fee.go::Mempoolhandler, L103 due to low fee

File: ante/fee.go (hunk 2 above)
Comment (psy2848048) [reply ante/fee.go:209]:
```
insufficient fees; got: 160000TKN required: 680000000000000000000TKN,6800000000000000axpla: insufficient fee
```

Errored math
Let me check and derive the correct math

Comment (JoowonYun) [reply ante/fee.go]: https://github.com/xpladev/xpla/pull/58/commits/cf5181ea4e7dc12e916e53698ed88d410f178d18

File: ante/fee.go [hunk 3]
```diff
@@ -91,7 +91,7 @@ func (mfd MempoolFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate b
 				if gp.Denom != xplatypes.DefaultDenom {
 					xatp, found := mfd.xatpKeeper.GetXatp(ctx, gp.Denom)
 					if found {
-						decimal := sdk.NewDecFromIntWithPrec(sdk.NewInt(10), int64(xatp.Decimals))
+						decimal := sdk.NewDecFromIntWithPrec(sdk.NewInt(1), int64(xatp.Decimals))
```
Comment (psy2848048) [inline ante/fee.go:94]:
```go
decimal := sdk.NewDecFromIntWithPrec(sdk.NewInt(1), xplatypes.DefaultDenomPrecision-int64(xatp.Decimals))
fee = fee.Ceil().Mul(decimal)
```

Comment (JoowonYun) [reply ante/fee.go]: https://github.com/xpladev/xpla/pull/58/commits/55c8d94cdfc4e5ec6525278db6e37f8854fba96b

File: ante/fee.go (hunk 1 above)
Comment (psy2848048) [reply ante/fee.go:218]: FYI

Also fails that the simulation does even not raise simulation error with XPLA fee
https://github.com/xpladev/xpla/blob/test/xatp/integration_test/integration_test.go#L788

Comment (JoowonYun) [reply ante/fee.go:218]: For simulation, Tx must be in sync mode. If you use sync mode you will get an error in TxResponse. It is not a broadcastTx error.

https://github.com/xpladev/xpla/blob/test/xatp/integration_test/wallet_test.go#L242-L250

Comment (psy2848048) [reply ante/fee.go:218]: Solved as below in the testing side and works well 👍 

```go
if currtestingenv := os.Getenv("GOLANG_TESTING"); currtestingenv != "true" {
res, err := client.BroadcastTx(context.Background(), &txtype.BroadcastTxRequest{
	TxBytes: txBytes,
	Mode:    mode,
})

if err != nil {
	err = errors.Wrap(err, "broadcastTx")
	return "", err
}

// if the tx works well in the sync mode, it returns as "[]"
if res.TxResponse.RawLog != "[]" {
	return "", fmt.Errorf(res.TxResponse.RawLog)
}
```

**Review by psy2848048 — APPROVED** [review]

**Pull request merged**